pkg/reconciler/tenancy/initialization/apibinder_initializer_controller.go

```diff
@@ -0,0 +1,204 @@
+	}
+
+	clusterWorkspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueClusterWorkspace(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueClusterWorkspace(obj)
+		},
+	})
+
+	return c, nil
+}
+
+// APIBinder is a controller which instantiates APIBindings and waits for them to be fully bound
+// in new ClusterWorkspaces.
+type APIBinder struct {
+	initializer tenancyv1alpha1.ClusterWorkspaceInitializer
+
+	kcpClusterClient kcpclient.ClusterInterface
+
+	clusterWorkspaceLister tenancylister.ClusterWorkspaceLister
+
+	queue workqueue.RateLimitingInterface
+}
+
+func (b *APIBinder) enqueueClusterWorkspace(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	// TODO(nrb): ADD LOGGER
+	b.queue.Add(key)
+}
+
+func (b *APIBinder) startWorker(ctx context.Context) {
+	for b.processNextWorkItem(ctx) {
+	}
+}
+
+func (b *APIBinder) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer b.queue.ShutDown()
+	logger := logging.WithReconciler(klog.FromContext(ctx), apiBinderControllerName)
+	ctx = klog.NewContext(ctx, logger)
+
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, b.startWorker, time.Second)
+	}
+	<-ctx.Done()
+}
+
+func (b *APIBinder) ShutDown() {
+	b.queue.ShutDown()
+}
+
+func (b *APIBinder) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := b.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer b.queue.Done(key)
+
+	if err := b.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%s: failed to sync %q, err: %w", apiBinderControllerName, key, err))
+		b.queue.AddRateLimited(key)
+		return true
+	}
+
+	b.queue.Forget(key)
+	return true
+}
+
+func (b *APIBinder) process(ctx context.Context, key string) error {
+	logger := klog.FromContext(ctx)
+	logger = logger.WithValues("clusterWorkspace", key)
+	clusterWorkspace, err := b.clusterWorkspaceLister.Get(key)
+
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			parentClusterName, clusterWorkspaceName := clusters.SplitClusterAwareKey(key)
+			logger.Error(err, "failed to get ClusterWorkspace from lister", "parentCluster", parentClusterName, "clusterWorkspace", clusterWorkspaceName)
+		}
+		return nil // nothing we can do here
+	}
+
+	updatedClusterWorkspace := clusterWorkspace.DeepCopy()
+	b.reconcile(ctx, updatedClusterWorkspace)
+
+	// If the object being reconciled changed as a result, update it.
+	return b.patchIfNeeded(ctx, clusterWorkspace, updatedClusterWorkspace)
+}
+
+func (b *APIBinder) patchIfNeeded(ctx context.Context, old, obj *tenancyv1alpha1.ClusterWorkspace) error {
```

Comment:
nit: we now have a generic library (`pkg/reconciler/comitter`) that you can use to not have to duplicate this code.